### PR TITLE
feat(vscode): read design token file patterns from Angular.json

### LINKED ISSUE #1489: [Feature]: Compute design token files patterns from angular.json
Would be nice to get this information from closest `angular.json` (or `nx.json` or `project.json`)

_Originally posted by @kpanot in https://github.com/AmadeusITGroup/otter/pull/1479#discussion_r1522391268_

### PR DESCRIPTION
**Pull request by kpanot**
## Proposed change

Read design token file patterns from Angular.json if not specified

## Related issues

- :rocket: Feature resolves #1489

### CODE REVIEW

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 1]
```diff
@@ -5,8 +5,40 @@ import * as path from 'node:path';
...
+        Object.values(node).forEach((value) => walkThroughAngularFile(value, mem));
+      }
+    }
+    return mem;
+  };
+
+  const patterns = (await Promise.all(angularFile
+    .map(async (file) => {
+      const contnent = JSON.parse((await vscode.workspace.fs.readFile(file)).toString());
+      return [...walkThroughAngularFile(contnent)];
```
Comment (fpaul-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:33]:
```suggestion
      const content = JSON.parse((await vscode.workspace.fs.readFile(file)).toString());
      return [...walkThroughAngularFile(content)];
```

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 2]
```diff
@@ -5,8 +5,40 @@ import * as path from 'node:path';
 
 const readFile = async (filePat: string) => (await vscode.workspace.fs.readFile(vscode.Uri.parse(filePat))).toString();
 
+const getfilePatternFromProjectConfiguration = async (): Promise<string[] | undefined> => {
```
Comment (fpaul-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:8]:
```suggestion
const getFilesPatternsFromProjectConfiguration = async (): Promise<string[] | undefined> => {
```

**Review by fpaul-1A — APPROVED** [review]

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 3]
```diff
@@ -5,8 +5,40 @@ import * as path from 'node:path';
 
 const readFile = async (filePat: string) => (await vscode.workspace.fs.readFile(vscode.Uri.parse(filePat))).toString();
 
+const getFilesPatternsFromProjectConfiguration = async (): Promise<string[] | undefined> => {
+  const angularFile = await vscode.workspace.findFiles('angular.json');
```
Comment (matthieu-crouzet) [inline apps/vscode-extension/src/intellisense/design-token.ts:9]: We can maybe add the support of nx by adding project.json as well no ?

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 4]
```diff
@@ -5,8 +5,40 @@ import * as path from 'node:path';
...
+  if (angularFile.length === 0) {
+    return;
+  }
+
+  const walkThroughAngularFile = (node: any, mem: Set<string> = new Set()) => {
+    if (typeof node === 'object' && node !== null && !Array.isArray(node)) {
+      if (node.executor === '@o3r/design:generate-css' || node.builder === '@o3r/design:generate-css') {
+        node.options?.designTokenFilePatterns?.forEach((pattern: string) => mem.add(pattern));
+        if (node.options?.configurations) {
```
Comment (cpaulve-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:18]: which configurations are you targeting? Isn't the configurations node at the same level as node.options?
image

Comment (kpanot) [reply apps/vscode-extension/src/intellisense/design-token.ts:18]: oops, I trusted too much copilot on this one :S

**Review by cpaulve-1A — APPROVED** [review]

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 5]
```diff
@@ -10,9 +10,45 @@ type DesignTokenCache = {
 
 const readFile = async (filePat: string) => (await vscode.workspace.fs.readFile(vscode.Uri.parse(filePat))).toString();
 
+const getFilesPatternsFromProjectConfiguration = async (currentFile: string): Promise<string[] | undefined> => {
+  const angularFile = vscode.workspace.findFiles('angular.json');
+  const nxFile = vscode.workspace.findFiles('nx.json');
+  const projectFiles = vscode.workspace.findFiles('**/project.json')
+    .then((files) => files.filter((file) => path.relative(path.dirname(file.fsPath), currentFile).includes('..')));
```
Comment (fpaul-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:17]: I am not sure to understand this part, do we want to check all `project.json` except the one related to the current file ?
or is a `!` missing ?

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 6]
```diff
@@ -10,9 +10,45 @@ type DesignTokenCache = {
...
+  const configFiles = (await Promise.all([angularFile, nxFile, projectFiles])).flat();
+  if (configFiles.length === 0) {
+    return;
+  }
+
+  const walkThroughAngularFile = (node: any, mem: Set<string> = new Set()) => {
```
Comment (cpaulve-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:23]: You could rename it to walkThroughProjectFiles

**Review by fpaul-1A — APPROVED** [review]

File: apps/vscode-extension/src/intellisense/design-token.ts [hunk 7]
```diff
@@ -10,9 +10,45 @@ type DesignTokenCache = {
 
 const readFile = async (filePat: string) => (await vscode.workspace.fs.readFile(vscode.Uri.parse(filePat))).toString();
 
+const getFilesPatternsFromProjectConfiguration = async (currentFile: string): Promise<string[] | undefined> => {
```
Comment (pginoux-1A) [inline apps/vscode-extension/src/intellisense/design-token.ts:13]: How long does the execution take on a big project ? Won't we risk to slow down the IDE ?

**Review by cpaulve-1A — APPROVED** [review]

**Pull request merged**